### PR TITLE
fix(ApplicationCommand): use new ApplicationCommandOptionType enums

### PR DESCRIPTION
**Pull request by Koyamie**
**Please describe the changes this PR makes and why it should be merged:**
This fixes the ApplicationCommandOptionType with new enums.

**Status and versioning classification:**

Please move lines that apply to you out of the comment:
- Code changes have been tested against the Discord API, or there are no code changes

### CODE REVIEW

**Review by vladfrangu — APPROVED** [review]
This is why we use Enums for this...

**Comment by Koyamie** [comment]
Fixed `ApplicationCommandOptionType` in `transformOption` too.

**Comment by Koyamie** [comment]
Fixed components that were not using discord api types.

**Comment by ImRodry** [comment]
> This is why we use Enums for this...

so why don't we change all of these instances in this PR to use enums? Would be quite easy

File: packages/discord.js/src/structures/ApplicationCommand.js [hunk 1]
```diff
@@ -379,7 +379,7 @@ class ApplicationCommand extends Base {
       name: option.name,
       description: option.description,
       required:
-        option.required ?? (stringType === 'SUB_COMMAND' || stringType === 'SUB_COMMAND_GROUP' ? undefined : false),
```
Comment (iCrawl) [inline packages/discord.js/src/structures/ApplicationCommand.js:382]: Revert this.

File: packages/discord.js/src/structures/CommandInteractionOptionResolver.js [hunk 2]
```diff
@@ -38,12 +38,12 @@ class CommandInteractionOptionResolver {
     this._hoistedOptions = options;
 
     // Hoist subcommand group if present
-    if (this._hoistedOptions[0]?.type === 'SUB_COMMAND_GROUP') {
+    if (this._hoistedOptions[0]?.type === 'SubcommandGroup') {
       this._group = this._hoistedOptions[0].name;
       this._hoistedOptions = this._hoistedOptions[0].options ?? [];
     }
     // Hoist subcommand if present
-    if (this._hoistedOptions[0]?.type === 'SUB_COMMAND') {
```
Comment (iCrawl) [inline packages/discord.js/src/structures/CommandInteractionOptionResolver.js:46]: Revert this.

File: packages/discord.js/src/structures/CommandInteractionOptionResolver.js [hunk 3]
```diff
@@ -145,7 +145,7 @@ class CommandInteractionOptionResolver {
    * The value of the option, or null if not set and not required.
    */
   getChannel(name, required = false) {
-    const option = this._getTypedOption(name, 'CHANNEL', ['channel'], required);
```
Comment (iCrawl) [inline packages/discord.js/src/structures/CommandInteractionOptionResolver.js:148]: Revert this.

File: packages/discord.js/src/structures/CommandInteractionOptionResolver.js [hunk 4]
```diff
@@ -133,7 +133,7 @@ class CommandInteractionOptionResolver {
    * @returns {?boolean} The value of the option, or null if not set and not required.
    */
   getBoolean(name, required = false) {
-    const option = this._getTypedOption(name, 'BOOLEAN', ['value'], required);
```
Comment (iCrawl) [inline packages/discord.js/src/structures/CommandInteractionOptionResolver.js:136]: Revert this.

File: packages/discord.js/src/structures/CommandInteractionOptionResolver.js [hunk 5]
```diff
@@ -167,7 +167,7 @@ class CommandInteractionOptionResolver {
    * @returns {?number} The value of the option, or null if not set and not required.
    */
   getInteger(name, required = false) {
-    const option = this._getTypedOption(name, 'INTEGER', ['value'], required);
```
Comment (iCrawl) [inline packages/discord.js/src/structures/CommandInteractionOptionResolver.js:170]: Revert this.

File: packages/discord.js/src/structures/CommandInteractionOptionResolver.js [hunk 6]
```diff
@@ -178,7 +178,7 @@ class CommandInteractionOptionResolver {
    * @returns {?number} The value of the option, or null if not set and not required.
    */
   getNumber(name, required = false) {
-    const option = this._getTypedOption(name, 'NUMBER', ['value'], required);
```
Comment (iCrawl) [inline packages/discord.js/src/structures/CommandInteractionOptionResolver.js:181]: Revert this.

File: packages/discord.js/src/structures/CommandInteractionOptionResolver.js [hunk 7]
```diff
@@ -189,7 +189,7 @@ class CommandInteractionOptionResolver {
    * @returns {?User} The value of the option, or null if not set and not required.
    */
   getUser(name, required = false) {
-    const option = this._getTypedOption(name, 'USER', ['user'], required);
```
Comment (iCrawl) [inline packages/discord.js/src/structures/CommandInteractionOptionResolver.js:192]: Revert this.

File: packages/discord.js/src/structures/CommandInteractionOptionResolver.js [hunk 8]
```diff
@@ -224,7 +224,7 @@ class CommandInteractionOptionResolver {
    * The value of the option, or null if not set and not required.
    */
   getMentionable(name, required = false) {
-    const option = this._getTypedOption(name, 'MENTIONABLE', ['user', 'member', 'role'], required);
```
Comment (iCrawl) [inline packages/discord.js/src/structures/CommandInteractionOptionResolver.js:227]: Revert this.

File: packages/discord.js/src/structures/CommandInteractionOptionResolver.js [hunk 9]
```diff
@@ -212,7 +212,7 @@ class CommandInteractionOptionResolver {
    * @returns {?(Role|APIRole)} The value of the option, or null if not set and not required.
    */
   getRole(name, required = false) {
-    const option = this._getTypedOption(name, 'ROLE', ['role'], required);
```
Comment (iCrawl) [inline packages/discord.js/src/structures/CommandInteractionOptionResolver.js:215]: Revert this.

File: packages/discord.js/src/structures/CommandInteractionOptionResolver.js [hunk 10]
```diff
@@ -201,7 +201,7 @@ class CommandInteractionOptionResolver {
    * The value of the option, or null if not set and not required.
    */
   getMember(name, required = false) {
-    const option = this._getTypedOption(name, 'USER', ['member'], required);
```
Comment (iCrawl) [inline packages/discord.js/src/structures/CommandInteractionOptionResolver.js:204]: Revert this.

File: packages/discord.js/src/structures/MessageActionRow.js [hunk 11]
```diff
@@ -40,7 +40,7 @@ class MessageActionRow extends BaseMessageComponent {
    * @param {Client} [client] The client constructing this MessageActionRow, if provided
    */
   constructor(data = {}, client = null) {
-    super({ type: 'ACTION_ROW' });
```
Comment (iCrawl) [inline packages/discord.js/src/structures/MessageActionRow.js:43]: Revert this.

File: packages/discord.js/src/structures/MessageButton.js [hunk 12]
```diff
@@ -24,7 +24,7 @@ class MessageButton extends BaseMessageComponent {
    * @param {MessageButton|MessageButtonOptions} [data={}] MessageButton to clone or raw data
    */
   constructor(data = {}) {
-    super({ type: 'BUTTON' });
```
Comment (iCrawl) [inline packages/discord.js/src/structures/MessageButton.js:27]: Revert this.

File: packages/discord.js/src/structures/MessageSelectMenu.js [hunk 13]
```diff
@@ -41,7 +41,7 @@ class MessageSelectMenu extends BaseMessageComponent {
    * @param {MessageSelectMenu|MessageSelectMenuOptions} [data={}] MessageSelectMenu to clone or raw data
    */
   constructor(data = {}) {
-    super({ type: 'SELECT_MENU' });
```
Comment (iCrawl) [inline packages/discord.js/src/structures/MessageSelectMenu.js:44]: Revert this.

**Review by iCrawl — CHANGES_REQUESTED** [review]

**Comment by Koyamie** [comment]
I should close the PR at this point lol

**Comment by ImRodry** [comment]
What’s the point of leaving all of that broken even if it’s going to be changed soon? The dev builds are unusable at the moment because of this

**Comment by iCrawl** [comment]
@Koyamie The other fixes are still valid

**Review by iCrawl — APPROVED** [review]

**Pull request merged**